prov/sm2/src/sm2_ep.c

```diff
@@ -271,22 +257,13 @@ static ssize_t sm2_do_inject(struct sm2_ep *ep, struct sm2_region *peer_smr,
 			     size_t iov_count, size_t total_len, void *context)
 {
 	struct sm2_xfer_entry *xfer_entry;
-	struct sm2_region *self_region;
-	struct sm2_av *av =
-		container_of(ep->util_ep.av, struct sm2_av, util_av);
-	struct sm2_mmap *map = &av->mmap;
+	ssize_t ret;
 
 	assert(total_len <= SM2_INJECT_SIZE);
 
-	self_region = sm2_mmap_ep_region(map, ep->gid);
-
-	if (smr_freestack_isempty(sm2_freestack(self_region))) {
-		sm2_progress_recv(ep);
-		if (smr_freestack_isempty(sm2_freestack(self_region)))
-			return -FI_EAGAIN;
-	}
-
-	xfer_entry = smr_freestack_pop(sm2_freestack(self_region));
+	ret = sm2_pop_xfer_entry(ep, &xfer_entry);
```

Comment:
We have the same code in CMA and IPC as well. I will add that to the CMA and IPC commits after you merge this.